vsscale/scale.py

```diff
@@ -516,3 +522,56 @@ class SwinUnetPhotoV2(BaseWaifu2x):
 
     class SwinUnetArtScan(BaseWaifu2x):
         _model = 10
+
+
+class SharpBilinear(ComplexScaler):
+    """
+    Pre-supersample using Point, then scale to the target dimensions with Bilinear.
+
+    This is used for pixel art to preserve the "sharpness", while still allowing for
+    non-integer ratio scaling without messing up the intended look of the image.
+    """
+
+    @inject_self.cached
+    @inject_kwargs_params
+    def scale(  # type: ignore[override]
+        self, clip: vs.VideoNode, width: int | None = None, height: int | None = None,
+        shift: tuple[TopShift, LeftShift] = (0, 0),
+        *,
+        border_handling: BorderHandling = BorderHandling.MIRROR,
+        sample_grid_model: SampleGridModel = SampleGridModel.MATCH_EDGES,
+        sar: Sar | bool | float | None = None, dar: Dar | bool | float | None = None, keep_ar: bool | None = None,
+        linear: bool | None = None, sigmoid: bool | tuple[Slope, Center] = False,
+        **kwargs: Any
+    ) -> vs.VideoNode:
+        target_width = width or clip.width
+        target_height = height or clip.height
+
+        # Ideally, we should always scale in linear light by default.
+        kernel_kwargs = dict(linear=fallback(linear, self.kwargs.get('linear', True)))
+
+        scale_kwargs = dict(
+            sar=sar, dar=dar, keep_ar=keep_ar,
+            sigmoid=sigmoid, border_handling=border_handling,
+            sample_grid_model=sample_grid_model
+        ) | dict(linear=kernel_kwargs.get('linear', False)) | kwargs
+
+        if (clip.width, clip.height) == (target_width, target_height):
+            return clip
+
+        if target_width <= clip.width and target_height <= clip.height:
+            return Bilinear.scale(clip, target_width, target_height, shift, **scale_kwargs)
+
+        max_ratio = max(target_width / clip.width, target_height / clip.height)
+        int_ratio = 2 ** int(log2(max_ratio) + 0.5)
```

Comment:
can I interest you in our newest invention, `math.ceil`

Reply:
I am die goodbye

Reply:
it’s transpired that this was, of course, `round`, not `math.ceil`, oops

Reply:
Actually, round isn't exactly the math round
Look here: https://realpython.com/python-rounding/#pythons-built-in-round-function

Reply:
Oh, so it’s round-half-to-even? I didn’t know that. Thanks!

Anyway, @LightArrowsEXE, why is this forced to be a power of two? And do you actually want to overshoot the ratio and then downscale with bilinear, undershoot and then upscale with bilinear, or pick the closest integer in any direction?

Reply:
Yes, exactly, it is round-half-to-even.

Reply:
@astiob I noticed while writing this that sometimes, it looks better when you downscale it vs. upscale it. I found that was the case if you had for example a factor of 2.5, upscaling looks better, but if you have for example a factor of 3.5, downscaling looks better.